File: FarNet/FarNetMan/AnyMenu.cpp

/*
FarNet plugin for Far Manager
Copyright (c) 2006-2013 Roman Kuzmin
*/

#include "StdAfx.h"
#include "AnyMenu.h"

namespace FarNet
{;
AnyMenu::AnyMenu()
: _x(-1)
, _y(-1)
, _selected(-1)
, _keyIndex(-1)
{
	_items = gcnew List<FarItem^>;
	WrapCursor = true; //! default is true, as recommended by Far API
}

int AnyMenu::X::get()
{
	return _x;
}

void AnyMenu::X::set(int value)
{
	_x = value;
}

int AnyMenu::Y::get()
{
	return _y;
}

void AnyMenu::Y::set(int value)
{
	_y = value;
}

Object^ AnyMenu::SelectedData::get()
{
	if (_selected < 0 || _selected >= _items->Count)
		return nullptr;
	return _items[_selected]->Data;
}

IList<FarItem^>^ AnyMenu::Items::get()
{
	return _items;
}

int AnyMenu::Selected::get()
{
	return _selected;
}

void AnyMenu::Selected::set(int value)
{
	_selected = value;
}

KeyData^ AnyMenu::Key::get()
{
	return _keyIndex < 0 ? KeyData::Empty : _keys[_keyIndex];
}

FarItem^ AnyMenu::Add(String^ text)
{
	return Add(text, nullptr);
}

FarItem^ AnyMenu::Add(String^ text, EventHandler<MenuEventArgs^>^ click)
{
	FarItem^ r = gcnew SetItem;
	r->Text = text;
	r->Click = click;
	Items->Add(r);
	return r;
}

void AnyMenu::AddKey(int virtualKeyCode)
{
	AddKey(virtualKeyCode, ControlKeyStates::None, nullptr);
}
void AnyMenu::AddKey(int virtualKeyCode, ControlKeyStates controlKeyState)
{
	AddKey(virtualKeyCode, controlKeyState, nullptr);
}
void AnyMenu::AddKey(int virtualKeyCode, ControlKeyStates controlKeyState, EventHandler<MenuEventArgs^>^ handler)
{
	_keys.Add(gcnew KeyData(virtualKeyCode, controlKeyState));
	_handlers.Add(handler);
}

}

File: FarNet/FarNet/Works/Dialog/AnyMenu.cs

// FarNet plugin for Far Manager
// Copyright (c) Roman Kuzmin

using System;
using System.Collections.Generic;

namespace FarNet.Works;
#pragma warning disable 1591

public abstract class AnyMenu : IAnyMenu
{
	protected readonly List<FarItem> myItems = [];

	protected readonly List<KeyData> myKeys = [];

	protected readonly List<EventHandler<MenuEventArgs>?> myHandlers = [];

	protected int myKeyIndex = -1;

	public bool AutoAssignHotkeys { get; set; }

	public bool NoShadow { get; set; }

	public bool SelectLast { get; set; }

	public bool ShowAmpersands { get; set; }

	public bool WrapCursor { get; set; }

	public int MaxHeight { get; set; }

	public int Selected { get; set; }

	public int X { get; set; }

	public int Y { get; set; }

	public IList<FarItem> Items { get { return myItems; } }

	public object? SelectedData { get { return Selected < 0 || Selected >= myItems.Count ? null : myItems[Selected].Data; } }

	public object? Sender { get; set; }

	public string? Bottom { get; set; }

	public string? HelpTopic { get; set; }

	public string? Title { get; set; }

	public abstract bool Show();

	public AnyMenu()
	{
		X = -1;
		Y = -1;
		Selected = -1;
		WrapCursor = true; //! default is true, as recommended by Far
	}

	public FarItem Add(string text)
	{
		return Add(text, null);
	}

	public FarItem Add(string text, EventHandler<MenuEventArgs>? click)
	{
		var r = new SetItem()
		{
			Text = text,
			Click = click
		};
		myItems.Add(r);
		return r;
	}

	public KeyData Key
	{
		get
		{
			return myKeyIndex < 0 ? KeyData.Empty : myKeys[myKeyIndex];
		}
	}

	public void AddKey(int virtualKeyCode)
	{
		AddKey(virtualKeyCode, ControlKeyStates.None, null);
	}

	public void AddKey(int virtualKeyCode, ControlKeyStates controlKeyState)
	{
		AddKey(virtualKeyCode, controlKeyState, null);
	}

	public void AddKey(int virtualKeyCode, ControlKeyStates controlKeyState, EventHandler<MenuEventArgs>? handler)
	{
		myKeys.Add(new KeyData(virtualKeyCode, controlKeyState));
		myHandlers.Add(handler);
	}
}
